---
title: "bam2IUPAC"
author: "Kristian K Ullrich"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{bam2IUPAC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

In this vignette users learn how to create IUPAC __fasta__ format files from __bam__ files which can be used from distIUPAC to calculate IUPAC distances.

## Prerequistes for obtaining __fasta__ format files from reference mapped __bam__ files

The following external tools needs to be installed to be able to obtain IUPAC __fasta__ format files:

1. Genome mapper of your choice (e.g. [http://bio-bwa.sourceforge.net/](http://bio-bwa.sourceforge.net/) or [https://github.com/Cibiv/NextGenMap][https://github.com/Cibiv/NextGenMap]))
2. Picard tools for __bam__ file sorting and de-duplication [http://broadinstitute.github.io/picard/](http://broadinstitute.github.io/picard/)
3. ANGSD for IUPAC __fasta__ format retrieval [https://github.com/ANGSD/angsd](https://github.com/ANGSD/angsd)

## Prerequisites installation - unix based systems (no MAC OS X)

Short description of how to compile the external tools for a unix based system are given.
See the next section for MAC OS X installation.

### NextGenMap installation
```
#download latest release from NextGenMap
wget https://codeload.github.com/Cibiv/NextGenMap/tar.gz/NextGenMap-0.5.5.tar.gz
tar -xvf v0.5.5.tar.gz
rm v0.5.5.tar.gz
cd NextGenMap-0.5.5/
mkdir -p build/
cd build/
cmake ..
make
```

### Picard tools installation
```
#download latest release 'picard.jar' from broadinstitute
wget https://github.com/broadinstitute/picard/releases/download/2.17.4/picard.jar
```

### ANGSD installation

Until now the functionality to also consider a certain nucleotide depth for IUPAC assignment is not directly implemented in ANGSD. However, below you will find all necessary changes that need to be included in the original code to get this functionality.

```
git clone https://github.com/samtools/htslib.git;
git clone https://github.com/angsd/angsd.git;
cd htslib;make;
#change the follwoing files from ANGSD as given below
cd ../angsd;
#
#'abcWriteFasta.h' CHANGES:
# Line 11 add:
# double iupacRatio;
#
#'abcWriteFasta.cpp' CHANGES:
# Line 29 add:
# fprintf(argFile,"\t-iupacRatio\t%.3f\t (Remove nucleotide below total depth ratio for IUPAC assignment)\n",iupacRatio);
#
# Line 230 change the code into:
#  }else if(doFasta==4){
#    //supplied by kristian ullrich
#    for(int s=0;s<pars->numSites&&pars->posi[s]<header->target_len[pars->refId];s++){
#      if(pars->keepSites[s]==0)
#        continue;
#      if(pars->nInd==1)
#        myFasta[pars->posi[s]] = intToIupac[ angsd::getIupacCount(pars->counts[s],0,iupacRatio) ];
#      else
#        myFasta[pars->posi[s]] = intToIupac[ angsd::getIupacCountTotal(pars->counts[s],pars->nInd,iupacRatio) ];
#    }
#  }
#
#'analysisFunction.h' CHANGES:
#
# Line 71 change the code into:
# int getIupacCount(suint *d, int i, double iRatio, int depth = -1);
#
# Line 74 change the code into:
# int getIupacCountTotal(suint *d, int nInd, double iRatio);
#
#'analysisFunction.cpp' CHANGES:
#
# Line 
# // iupack code supplied by kristian ullrich
# int angsd::getIupacCount(suint *counts,int i, double iRatio, int depth){
#
#  if(depth==-1){
#    depth=0;
#    for( int b = 0; b < 4; b++ )
#      depth+=counts[b+4*i];
#  }
#
#  if(depth<=0)
#    return 14;
#
#  int whichIUPAC = 0;
#  double bIUPACscore = 0;
#  for(int b=0;b<4;b++){
#    //if (counts[b+4*i]>0){
#    if (double(counts[b+4*i])/double(depth)>iRatio){
#      bIUPACscore = bIUPACscore + pow(b+1,2);
#    }
#  }
make HTSSRC=../htslib
```

## Prerequisites installation - MAC OS X systems

For MAC OS X systems there are additional prerequisites that needs to be installed to be able to compile all necessary software.

### Xcode Command Line Tools
'Xcode' needs to be installed from App-Store

### Homebrew
'Homebrew' needs to be installed [https://brew.sh/index_de.html](https://brew.sh/index_de.html)

```
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)
```

### Git
'Git' needs to be installe [https://git-scm.com/download/mac](https://git-scm.com/download/mac)

1. download latest Git DMG Image
2. Open DMG Iamge and Install PKG file

### Autoconf via Homebrew
'Autoconf' needs to be installed [http://www.gnu.org/software/autoconf/autoconf.html](http://www.gnu.org/software/autoconf/autoconf.html)
```
brew install autoconf
```

### Cmake
'Cmake' needs to be installed [https://cmake.org/download/](https://cmake.org/download/)

1. download latest cmake DMG Image release [https://cmake.org/files/v3.10/cmake-3.10.2-Darwin-x86_64.dmg](https://cmake.org/files/v3.10/cmake-3.10.2-Darwin-x86_64.dmg)
2. Open DMG Image and copy to Applications
3. in a Terminal
```
sudo "/Applications/CMake.app/Contents/bin/cmake-gui" --install
``

### NextGenMap installation - MAC OS X systems
```
#download latest release from NextGenMap
curl -L https://github.com/Cibiv/NextGenMap/archive/v0.5.5.tar.gz > v0.5.5.tar.gz
tar -xvf v0.5.5.tar.gz
rm v0.5.5.tar.gz
cd NextGenMap-0.5.5/
mkdir -p build/
cd build/
cmake ..
make
```

### Picard tools installation - MAC OS X systems
```
#download latest release 'picard.jar' from broadinstitute
curl -L https://github.com/broadinstitute/picard/releases/download/2.17.4/picard.jar > picard.jar
```

### ANGSD installation

```
git clone https://github.com/samtools/htslib.git;
git clone https://github.com/angsd/angsd.git;
cd htslib;/usr/local/Cellar/autoconf/2.69/bin/autoconf;/usr/local/Cellar/autoconf/2.69/bin/autoheader;./configure --disable-lzma;make;
#change the follwoing files from ANGSD as given below
cd ../angsd;
#
#'abcWriteFasta.h' CHANGES:
# Line 11 add:
# double iupacRatio;
#
#'abcWriteFasta.cpp' CHANGES:
# Line 29 add:
# fprintf(argFile,"\t-iupacRatio\t%.3f\t (Remove nucleotide below total depth ratio for IUPAC assignment)\n",iupacRatio);
#
# Line 230 change the code into:
#  }else if(doFasta==4){
#    //supplied by kristian ullrich
#    for(int s=0;s<pars->numSites&&pars->posi[s]<header->target_len[pars->refId];s++){
#      if(pars->keepSites[s]==0)
#        continue;
#      if(pars->nInd==1)
#        myFasta[pars->posi[s]] = intToIupac[ angsd::getIupacCount(pars->counts[s],0,iupacRatio) ];
#      else
#        myFasta[pars->posi[s]] = intToIupac[ angsd::getIupacCountTotal(pars->counts[s],pars->nInd,iupacRatio) ];
#    }
#  }
#
#'analysisFunction.h' CHANGES:
#
# Line 71 change the code into:
# int getIupacCount(suint *d, int i, double iRatio, int depth = -1);
#
# Line 74 change the code into:
# int getIupacCountTotal(suint *d, int nInd, double iRatio);
#
#'analysisFunction.cpp' CHANGES:
#
# Line 
# // iupack code supplied by kristian ullrich
# int angsd::getIupacCount(suint *counts,int i, double iRatio, int depth){
#
#  if(depth==-1){
#    depth=0;
#    for( int b = 0; b < 4; b++ )
#      depth+=counts[b+4*i];
#  }
#
#  if(depth<=0)
#    return 14;
#
#  int whichIUPAC = 0;
#  double bIUPACscore = 0;
#  for(int b=0;b<4;b++){
#    //if (counts[b+4*i]>0){
#    if (double(counts[b+4*i])/double(depth)>iRatio){
#      bIUPACscore = bIUPACscore + pow(b+1,2);
#    }
#  }
make HTSSRC=../htslib
```

## STEP 1: Mapping __fastq__ files to a reference

### Using __BWA__ for mapping:

The user needs:
1. __fastq__ files, prefernetially QC preprocessed (see [http://www.usadellab.org/cms/?page=trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) for one of many trimming options of __fastq__ files)
2. reference __fasta__ file
3. build index for the reference
4. map reads
5. sort reads

```
#build index
bwa index ref.fasta
#map reads
#set the correct number of available threads with '-t' and add the corresponding read group tags for your sample
#here, 12 threads were used and the sample is named SAMPLE1 with paired-end data
bwa mem -t 12 -M -R @RG\tID:SAMPLE1\tLB:lib1\tPL:ILLUMINA\tPU:unit1\tSM:SAMPLE1 -o sample2ref.sam ref.fasta SAMPLE1.fwd.fq SAMPLE1.rev.fq
#sort sam file
java -jar picard.jar SortSam I=sample2ref.sam O=sample2ref.sorted.bam SO=coordinate
```

### Using __NextGenMap__ for mapping

The usere needs:
1. __fastq__ files, prefernetially QC preprocessed (see [http://www.usadellab.org/cms/?page=trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) for one of many trimming options of __fastq__ files)
2. reference __fasta__ file
3. map reads
4. sort reads

```
#map reads
#set the correct number of available threads with '-t' and add the corresponding read group tags for your sample
#here, 12 threads were used and the sample is named SAMPLE1 with paired-end data
ngm -1 SAMPLE1.fwd.fq -2 SAMPLE1.rev.fq -r ref.fasta -o sample2ref.bam --no-unal --sensitive -t 12 --no-progress --rg-id SAMPLE1 --rg-sm SAMPLE1 --rg-lb lib1 --rg-pl illumina --rg-pu unit1 -b
#sort bam file
java -jar picard.jar SortSam I=sample2ref.bam O=sample2ref.sorted.bam SO=coordinate
```
## OPTIONAL STEP: Merge sorted __bam__ files

```
java -jar picard.jar MergeSamFiles I=file1.sorted.bam I=file2.sorted.bam O=merged.sorted.bam
```

## STEP 2: Remove duplicates from sorted __bam__ file

```
java -jar picard.jar RemoveDuplicates REMOVE_DUPLICATES=true I=sample2ref.sorted.bam O=sample2ref.sorted.nodup.bam M=sample2ref.sorted.bam.duplicate.metrics
```

## STEP 3: Create IUPAC __fasta__ file for the de-duplicated referenced mapped __bam__ file

Chromosomes should be processed separately to be able to merge different samples into chromosome alignments, which can be processed with __distIUPAC__.

```
#chromosomes should be processed separately to be able to easily merge different samples into one alignment to be processed with 'distIUPAC'
#chromosome 'chr1' will be processed here
angsd -doFasta 4 -doCounts 1 -minQ 20 -minMapQ 30 -uniqueOnly -setMinDepth 5 -setMaxDepth 1000 -iupacRatio 0.2 -i sample2ref.sorted.nodup.bam -out SAMPLE1.minQ20.minMapQ30.guniqueOnly.setMinDepth5.setMaxDepth100.chr1 -r chr1
```

## STEP 4: Repeat STEP 1 to STEP 3 for multiples samples

## STEP 5: Merge samples into chromosome alignments

Assuming all IUPAC __fasta__ files are located in the same folder and same chromosome files have the same ending

```
#chromosome 'chr1' from different samples will be processed here
#1. uncompress fasta files
for file in *.chr1.fa.gz;do gunzip $file;done
#2. rename fasta sequences according to file names
for file in *.chr1.fa;do sed -i 's/>chr1/>'"$file"'/g' $file;done
#3. merge fasta files
for file in *.chr1.fa;do cat $file >> chr1.fa;done
```
## Calculating IUPAC distances for 'chr1' with __distIUPAC__

The following commands needs to be executed in __R__

```
library(distIUPAC)
dna<-readBStringSet("chr1.fa")
chr1.dist<-distIUPAC(as.character(subseq(dna,1,10000)))
```

## References

Korneliussen TS, Albrechtsen A and Nielsen R. __ANGSD: Analysis of Next Generation Sequencing Data.__ _BMC Bioinformatics_ (2014) _15_:356 [https://doi.org/10.1186/s12859-014-0356-4](https://doi.org/10.1186/s12859-014-0356-4)

Li, H. & Durbin, R. __Fast and accurate short read alignment with Burrows-Wheeler transform.__ _Bioinformatics_ (2009) _25_:1754 [https://doi.org/10.1093/bioinformatics/btp324](https://doi.org/10.1093/bioinformatics/btp324)

Picard tools [http://broadinstitute.github.io/picard](http://broadinstitute.github.io/picard)

Sedlazeck FJ, Rescheneder P, von Haeseler A. __NextGenMap: fast and accurate read mapping in highly polymorphic genomes.__ _Bioinformatics_ (2013) _21_:2790 [https://doi.org/10.1093/bioinformatics/btt468](https://doi.org/10.1093/bioinformatics/btt468)
