<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Pre-requistes for obtaining <strong>fasta</strong> format files from reference mapped <strong>bam</strong> files</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>





<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p>In this vignette Users learn how to use <a href="https://github.com/kullrich/distIUPAC">distIUPAC</a> to calculate <strong><code>IUPAC distances</code>__ based on __<code>IUPAC fasta format</code></strong> (<a href="https://www.bioinformatics.org/sms/iupac.html">IUPAC code</a>) files.</p>

<p>Next to learn how to calculate a distance matrix for one sequence alignment also within population distances (<strong><code>xStats</code><strong>) and between populations distances (</strong><code>xyStats</code><strong>, __<code>xyMultiStats</code></strong> and <code>xyzStats</code></strong>) based on sliding windows will be shown.</p>

<h2>Pre-requistes for obtaining <strong>fasta</strong> format files from reference mapped <strong>bam</strong> files</h2>

<p>One pre-requistes is a <strong>fasta</strong> format alignment file (same length for all sequences) either in <strong><code>non IUPAC fasta format</code>__ (&#39;ACTG-N&#39;) or in __<code>IUPAC fasta format</code></strong>_ (&#39;ACGTRYSWKMBDHV.-NX&#39;).</p>

<p>To get <strong><code>IUPAC fasta format</code>__ files from __<code>reference mapped bam</code></strong> files see the vignette <a href="https://github.com/kullrich/distIUPAC/tree/master/vignettes/bam2IUPAC.Rmd">01. bam2IUPAC vignette</a>.</p>

<h2>STEP 1: Loading</h2>

<h3>Using <strong><code>BWA</code></strong> for mapping:</h3>

<p>The user needs to perform five steps:</p>

<ol>
<li><strong><code>fastq</code>__ files, prefernetially __<code>QC</code></strong> pre-processed (see <a href="http://www.usadellab.org/cms/?page=trimmomatic">trimmomatic</a> for one of many trimming tools available)</li>
<li>reference <strong><code>fasta</code></strong> file</li>
<li><strong><code>build index</code></strong> for the reference</li>
<li><strong><code>map reads</code></strong></li>
<li><strong><code>sort reads</code></strong></li>
</ol>

<pre><code>#FORWARD FASTQ FILE: 1.fq
#REVERSE FASTQ FILE: 2.fq
#REFERENCE: ref.fasta
#USED THREADS: 12
#SAMPLE NAME: IND1
#SAM OUTPUT FILE: ind1.sam
#build index
bwa index ref.fasta
#map reads
bwa mem -t 12 -M -R @RG\tID:IND1\tLB:lib1\tPL:ILLUMINA\tPU:unit1\tSM:IND1 -o ind1.sam ref.fasta 1.fq 2.fq
#sort sam file
java -jar picard.jar SortSam I=ind1.sam O=ind1.sorted.bam SO=coordinate
</code></pre>

<h3>Using <strong><code>NextGenMap</code></strong> for mapping</h3>

<p>The usere needs to perform four steps:</p>

<ol>
<li><strong><code>fastq</code>__ files, prefernetially __<code>QC</code></strong> pre-processed (see <a href="http://www.usadellab.org/cms/?page=trimmomatic">trimmomatic</a> for one of many trimming tools available)</li>
<li>reference <strong><code>fasta</code></strong> file</li>
<li><strong><code>map reads</code></strong></li>
<li><strong><code>sort reads</code></strong></li>
</ol>

<pre><code>#FORWARD FASTQ FILE: 1.fq
#REVERSE FASTQ FILE: 2.fq
#REFERENCE: ref.fasta
#USED THREADS: 12
#SAMPLE NAME: IND1
#SAM OUTPUT FILE: ind1.bam
#map reads
ngm -1 1.fq -2 2.fq -r ref.fasta -o ind1.bam --no-unal --sensitive -t 12 --no-progress --rg-id IND1 --rg-sm IND1 --rg-lb lib1 --rg-pl illumina --rg-pu unit1 -b
#sort bam file
java -jar picard.jar SortSam I=ind1.bam O=ind1.sorted.bam SO=coordinate
</code></pre>

<h2>OPTIONAL STEP: Merge <strong><code>sorted bam</code></strong> files</h2>

<p>In case that multiple <strong><code>fastq</code>__ libraries exists or have been mapped without prior merging of the __<code>fastq</code>__ files, the User can merge __<code>reference mapped bam</code></strong> files as follows:</p>

<pre><code>java -jar picard.jar MergeSamFiles I=ind1_1.sorted.bam I=ind1_2.sorted.bam O=ind1.sorted.bam
</code></pre>

<h2>STEP 2: Remove duplicates from <strong><code>sorted bam</code></strong> file</h2>

<pre><code>java -jar picard.jar RemoveDuplicates REMOVE_DUPLICATES=true I=ind1.sorted.bam O=ind1.sorted.nodup.bam M=ind1.sorted.bam.duplicate.metrics
</code></pre>

<h2>STEP 3: Create <strong><code>IUPAC fasta</code>__ file for the de-duplicated __<code>referenced mapped bam</code></strong> file</h2>

<p>Chromosomes should be processed separately to be able to merge different samples into <strong><code>chromosome alignments</code>_<em>, which can be processed with _</em><code>distIUPAC</code></strong> as follows:</p>

<pre><code>#chromosomes should be processed separately to be able to easily merge different samples into one alignment to be processed with &#39;distIUPAC&#39;
#chromosome &#39;chr1&#39; will be processed here
samtools index ind1.sorted.nodup.bam
angsd -doFasta 4 -doCounts 1 -minQ 20 -minMapQ 30 -uniqueOnly -setMinDepth 5 -setMaxDepth 100 -iupacRatio 0.2 -i ind1.sorted.nodup.bam -out IND1.minQ20.minMapQ30.uniqueOnly.setMinDepth5.setMaxDepth100.chr1 -r chr1
</code></pre>

<h2>STEP 4: Repeat STEP 1 to STEP 3 for multiple individuals</h2>

<h2>STEP 5: Merge samples into chromosome alignments</h2>

<p>Assuming all <strong><code>IUPAC fasta</code></strong> files are located in the same folder and same chromosome files have the same ending, individuals can be merged as follows:</p>

<pre><code>#chromosome &#39;chr1&#39; from different individuals will be processed here
#1. uncompress fasta files
for file in *.chr1.fa.gz;do gunzip $file;done
#2. rename fasta sequences according to file names
for file in *.chr1.fa;do sed -i &#39;s/&gt;chr1/&gt;&#39;&quot;$file&quot;&#39;/g&#39; $file;done
#3. merge fasta files
for file in *.chr1.fa;do cat $file &gt;&gt; chr1.fa;done
</code></pre>

<h2>Calculating <strong><code>IUPAC distances</code>__ for &#39;chr1&#39; with __<code>distIUPAC</code></strong></h2>

<p>The following commands needs to be executed in <strong><code>R</code></strong>:</p>

<pre><code>library(distIUPAC)
dna&lt;-readDNAStringSet(&quot;chr1.fa&quot;)
chr1.dist&lt;-distIUPAC(as.character(subseq(dna,1,10000)))
</code></pre>

<h2>References</h2>

<p>Korneliussen TS, Albrechtsen A and Nielsen R. <strong>ANGSD: Analysis of Next Generation Sequencing Data.</strong> <em>BMC Bioinformatics</em> (2014) <em>15</em>:356 <a href="https://doi.org/10.1186/s12859-014-0356-4">https://doi.org/10.1186/s12859-014-0356-4</a></p>

<p>Li, H. &amp; Durbin, R. <strong>Fast and accurate short read alignment with Burrows-Wheeler transform.</strong> <em>Bioinformatics</em> (2009) <em>25</em>:1754 <a href="https://doi.org/10.1093/bioinformatics/btp324">https://doi.org/10.1093/bioinformatics/btp324</a></p>

<p><strong>Picard tools</strong> <a href="http://broadinstitute.github.io/picard">http://broadinstitute.github.io/picard</a></p>

<p>Sedlazeck FJ, Rescheneder P, von Haeseler A. <strong>NextGenMap: fast and accurate read mapping in highly polymorphic genomes.</strong> <em>Bioinformatics</em> (2013) <em>21</em>:2790 <a href="https://doi.org/10.1093/bioinformatics/btt468">https://doi.org/10.1093/bioinformatics/btt468</a></p>

</body>

</html>
